proposals/0010-resource-element-type-validation.md

```diff
@@ -31,74 +38,74 @@ RETs for TypedBuffer variants may include:
-Texture resources conform to the rules for TypedBuffers.
-If the target is SPIR-V, then only the set of rules for RawBuffers apply. TypedBuffers
-like `StructuredBuffer` may have RETs that exceed 128 bits, for example, if the target 
-IR is SPIR-V. The TypedBuffer rules above are only enforced when the IR target is DXIL.
+Texture resources conform to the rules for typed buffers.
+If the target is SPIR-V, then only the set of rules for raw buffers apply. Typed buffers
+like `RWBuffer` may have RETs that exceed 128 bits, for example, if the target 
+IR is SPIR-V. The typed buffer rules above are only enforced when the IR target is DXIL.
 
 If someone writes `RWBuffer<MyCustomType>` and MyCustomType is not a 
 valid RET, there should be infrastructure to reject this RET and emit a message 
 explaining why it was rejected as an RET.
 
 ## Motivation
 Currently, there is an allow list of valid RETs. It must be modified with respect 
-to this spec. Anything that is not an int, uint, nor a floating-point type, or vectors 
-or matrices containing the aforementioned types, will be rejected. The allow list isn't
+to this spec. Anything that is not a valid RET will be rejected. The allow list isn't
 broad enough, because it doesn't include the case where the RET is user-defined. 
 Ideally, a user should be able to determine how any user-defined structure is invalid 
 as an RET. Some system should be in place to more completely enforce the rules for 
 valid and invalid RETs, as well as provide useful information on why they are invalid.
 
 For example, when targeting DXIL IR, `RWBuffer<double4> b : register(u4);` will emit
 an error in DXC, but will not in clang-dxc, despite the fact that `double4` is an 
-invalid RET for TypedBuffers.
+invalid RET for typed buffers.
 
 ## Proposed solution
 
 The proposed solution is to use some type_traits defined in the std library, create
 some custom builtins to use as type_traits, and join them together to define a 
 set of conceptual constraints for any RET that is used. These conceptual constraints
-will be applied to every TypedBuffer resource type that is defined, so that all
-TypedBuffer HLSL resources have the same rules about which RETs are valid. 
+will be applied to every typed buffer resource type that is defined, so that all
+typed buffer HLSL resources have the same rules about which RETs are valid. 
 Validation will occur upon resource type instantiation. Additionally, certain 
-resource types are RawBuffer variants, such as `StructuredBuffer`. These
-resource types will have a different set of type-traits applied, which will
-loosen constraints on viable RETs.
+resource types are raw buffer variants, such as `StructuredBuffer`. These resource 
+types will have a different set of type-traits applied, which will loosen constraints
+on viable RETs. The type traits will be validated as `concepts`, which is a feature 
+introduced in C++20 to statically validate a type.

```

Comment:
Are we really planning on adding concepts to HLSL?

Reply:
Further discussion is necessary, though it seems easy to add to HLSL. However, while resource definitions are in HLSLExternalSemaSource.cpp, a C++ file, we can use concepts along with the resource definitions. If we plan to move resource definitions to HLSL files, we will either need to add concepts to HLSL, or implement RET validation some other way. The spec as it stands now specifies a design while resources are still defined in a C++ file though.